File: c_ds/train/basics/movingAverage.cpp
#include <deque>
#include <iostream>
#include <vector>
using namespace std;
class MovingAverage {
    std::deque<int> m_queue;
    int             m_avg_size;
    long long       m_sum;
public:
    MovingAverage(int size):m_queue(), m_avg_size(size), m_sum(0) { }
    
    double next(int val) {
        m_queue.push_back(val);
        m_sum += val;
        auto count = m_queue.size();
        while (count > m_avg_size){
            auto val_ = m_queue.front();
            m_sum -= val_;
            m_queue.pop_front();
            count = m_queue.size();
        }
        double result = double(m_sum) / double(count);
        return result; 
    }
};

/**
 * Your MovingAverage object will be instantiated and called as such:
 * MovingAverage* obj = new MovingAverage(size);
 * double param_1 = obj->next(val);
 */

int main(){
    MovingAverage ma(3);
    vector<int> inputs{1, 10, 3, 5};
    for ( auto val : inputs){
        cout << ma.next(val) << endl;
    }
    return 0;
}

File: c_ds/train/basics/movingAverage.py
from collections import deque
class MovingAverage:

    def __init__(self, size: int):
        self.__count__ = size
        self.__sum__ = 0
        self.__collection__ = deque()

    def next(self, val: int) -> float:
        self.__sum__ += val
        self.__collection__.append(val)
        count = len(self.__collection__)
        while count > self.__count__:
            del_val = self.__collection__.popleft()
            self.__sum__ -= del_val
            count -= 1
        return self.__sum__ / count
        


# Your MovingAverage object will be instantiated and called as such:
# obj = MovingAverage(size)
# param_1 = obj.next(val)
def main():
    ma = MovingAverage(3)
    inputs = [1, 10, 3, 5]
    for  val in inputs:
        print( ma.next(val))

if __name__ == "__main__":
    main()